File: packages/ce-core/src/stream.ts
import type { Budget, ContextItem, PackOptions } from "./types.js";
import { validatePackInputs } from "./schemas.js";
import { createScorer, defaultItemScorer } from "./score.js";
import { estimateTokens } from "./estimate.js";

/**
 * Stream-pack context items, yielding each selected item as it's chosen.
 *
 * Same greedy algorithm as pack() but yields items one at a time via
 * async generator. Useful for large item sets where you want to start
 * processing selected items before packing completes.
 *
 * Supports compression when `allowCompression` is set in options:
 * if an item exceeds the remaining budget, its compressions are tried
 * (largest fitting first), and a summarizer is consulted as a fallback.
 *
 * @param items - Context items to pack
 * @param budget - Token budget
 * @param options - Packing options (including allowCompression, summarizer)
 * @yields Selected ContextItems in score order
 * @throws {ValidationError} If items or budget fail validation
 * @throws {BudgetExceededError} If reserveTokens >= maxTokens
 *
 * @example
 * ```ts
 * for await (const item of packStream(items, { maxTokens: 4096 })) {
 *   console.log(`Selected: ${item.id}`);
 * }
 * ```
 */
export async function* packStream(
  items: ContextItem[],
  budget: Budget,
  options: PackOptions = {}
): AsyncGenerator<ContextItem> {
  validatePackInputs(items, budget);

  const scorer =
    options.scorer ??
    (options.weights ? createScorer(options.weights) : defaultItemScorer);
  const tokenEstimator = options.tokenEstimator;
  const maxTokens = budget.maxTokens - (budget.reserveTokens ?? 0);

  const scoredItems = items.map(item => {
    const tokens =
      item.tokens ??
      estimateTokens(item.content, { estimator: tokenEstimator });
    const score = scorer({ ...item, tokens });
    return { ...item, tokens, score };
  });

  const sorted = [...scoredItems].sort((a, b) => {
    if ((b.score ?? 0) === (a.score ?? 0)) {
      return (b.recency ?? 0) - (a.recency ?? 0);
    }
    return (b.score ?? 0) - (a.score ?? 0);
  });

  let remaining = Math.max(0, maxTokens);

  for (const item of sorted) {
    if ((item.tokens ?? 0) <= remaining) {
      remaining -= item.tokens ?? 0;
      yield item;
      continue;
    }

    // Attempt compression if enabled (C3 fix: match pack() behavior)
    if (options.allowCompression) {
      const compressed = tryCompress(item, remaining, options);
      if (compressed) {
        remaining -= compressed.tokens ?? 0;
        yield compressed;
      }
    }
  }
}

/**
 * Try to compress an item to fit within remainingTokens.
 * Picks the largest compression that fits (best quality).
 */
function tryCompress(
  item: ContextItem,
  remainingTokens: number,
  options: PackOptions
): ContextItem | null {
  const compressions = item.compressions ?? [];

  // Resolve tokens and sort descending (largest first = best quality)
  const withTokens = compressions.map(c => ({
    ...c,
    tokens:
      c.tokens ??
      estimateTokens(c.content, { estimator: options.tokenEstimator }),
  }));
  const sorted = withTokens.sort((a, b) => b.tokens - a.tokens);

  for (const compression of sorted) {
    if (compression.tokens <= remainingTokens) {
      return {
        ...item,
        content: compression.content,
        tokens: compression.tokens,
        metadata: {
          ...item.metadata,
          compressionNote: compression.note ?? "compression",
        },
      };
    }
  }

  // Fall back to summarizer
  if (options.summarizer) {
    const summary = options.summarizer(item, remainingTokens);
    if (summary) {
      const tokens =
        summary.tokens ??
        estimateTokens(summary.content, { estimator: options.tokenEstimator });
      if (tokens <= remainingTokens) {
        return { ...summary, tokens };
      }
    }
  }

  return null;
}

File: python/context_engineering/stream.py
"""Stream module: async generator variant of pack.

Yields items one at a time as they are selected, useful for large item sets.
"""
from __future__ import annotations

from typing import AsyncGenerator, List, Optional

from .core import Budget, ContextItem, estimate_tokens, calculate_weighted_score, ScoringWeights
from .errors import BudgetExceededError, ValidationError


async def pack_stream(
    items: List[ContextItem],
    budget: Budget,
    weights: Optional[ScoringWeights] = None,
    provider: Optional[str] = None,
) -> AsyncGenerator[ContextItem, None]:
    """Stream-pack context items, yielding each selected item as chosen.

    Same greedy algorithm as pack() but yields items one at a time via
    async generator. Useful for large item sets where you want to start
    processing selected items before packing completes.

    Args:
        items: Context items to pack.
        budget: Token budget.
        weights: Optional scoring weights.
        provider: Optional provider for token estimation.

    Yields:
        Selected ContextItems in score order.

    Raises:
        ValidationError: If budget.maxTokens <= 0.
        BudgetExceededError: If reserveTokens >= maxTokens.
    """
    if budget.max_tokens <= 0:
        raise ValidationError(
            f"maxTokens must be positive, got {budget.max_tokens}",
            [{"path": "maxTokens", "message": "must be positive"}],
        )
    if budget.reserve_tokens is not None and budget.reserve_tokens >= budget.max_tokens:
        raise BudgetExceededError(
            f"reserveTokens ({budget.reserve_tokens}) must be less than "
            f"maxTokens ({budget.max_tokens})"
        )

    max_tokens = budget.max_tokens - (budget.reserve_tokens or 0)

    scored = []
    for item in items:
        tokens = item.tokens or estimate_tokens(item.content, provider=provider)
        score = calculate_weighted_score(item, weights)
        scored.append(item.model_copy(update={"tokens": tokens, "score": score}))

    scored.sort(key=lambda x: (x.score or 0, x.recency or 0), reverse=True)

    remaining = max(0, max_tokens)

    for item in scored:
        if (item.tokens or 0) <= remaining:
            remaining -= item.tokens or 0
            yield item
